Add more developer guidelines

prajakta-gokhale opened the pull request:
Contributing back development guidelines created by OSRF and Amazon engineers.

Signed-off-by: Prajakta Gokhale <prajaktg@amazon.com>

wjwwood commented on source/Contributing.rst:
```diff
@@ -6,6 +6,26 @@ Contributing to ROS 2
+------
+* Respect what came before
+
+  ROS has been around for more than a decade and is used by developers and across the world. Keep a humble attitude and an open mindset while contributing.
```
nitpick: one sentence per line

See: https://index.ros.org/doc/ros2/Contributing/Developer-Guide/#markdown-restructured-text-docblocks

wjwwood commented on source/Contributing/Developer-Guide.rst:
```diff
@@ -54,13 +54,18 @@ Pull requests
   Separate changes should go into separate pull requests.
   See `GitHub's guide to writing the perfect pull request <https://github.com/blog/1943-how-to-write-the-perfect-pull-request>`__.
 
+  * As a rule of thumb, a pull request should avoid modifying more than 200 lines at once and must not modify more than 700 lines at once
```
That seems like a well intentioned, but arbitrary, restriction. There are plenty of cases where I've had pr's over 700 lines that would have been harder to test and review had they been broken into parts.

wjwwood replied:
nitpick: use a complete sentence here (since the other list items are currently sentences) and therefore use punctuation

wjwwood commented on source/Contributing/Developer-Guide.rst:
```diff
@@ -54,13 +54,18 @@ Pull requests
   (If you don't have access to the Jenkins job someone will trigger the jobs for you.)
+
+* A pull request must ideally contain a single commit.
+  Feedback can be integrated by amending the commit.
```
I don't think amending commits by default is the right approach. We smash all our pr's before commit (unless we explicitly want more than one commit), so having fixup commits does not harm readability in master, and helps to not break inline comments and preserve the history of iteration in pr's.

wjwwood commented on source/Contributing/Developer-Guide.rst:
```diff
@@ -142,6 +149,124 @@ C++ specific
+Software Development Lifecycle
+------------------------------
+
+This section describes step-by-step how to plan, design and implement a new feature:
```
nitpick: please include the "oxford comma" to avoid ambiguity and because we tend to use it elsewhere in our docs (in my experience), see: https://en.oxforddictionaries.com/explore/what-is-the-oxford-comma

wjwwood commented on source/Contributing/Developer-Guide.rst:
```diff
@@ -142,6 +149,124 @@ C++ specific
+
+* An issue should be created in the `ros2/ros2 <https://github.com/ros2/ros2/issues>`__ org, clearly describing the task being worked on.
+
+  * It should have a clear success criteria and highlight the concrete improvements expect from it.
```
`expect` -> `expected`

thomas-moulard commented:
@juanrh here is the PR we talked about - would be great if you could take a look at it!

thomas-moulard commented on source/Contributing.rst:
```diff
@@ -6,6 +6,26 @@ Contributing to ROS 2
+
+* Embed OSRF as early as possible
+
+  OSRF acts as a gate-keeper and advocate for the ROS community. Rely on their expertise and technical judgement from the design phase. Explicit approval from OSRF is required for any non-trivial work.
```
We can drop `Explicit approval from OSRF is required for any non-trivial work.` as this one is more an internal tenet (i.e. we don't expect everyone to do this).

Maybe let's add:

```
Start the discussion with OSRF and the community early. Long time ROS contributors may have a clearer vision of the bigger picture. If you implement a feature and send a PR without discussing with the community first, you are taking the risk to see it rejected or you may be asked to largely rethink your design. Opening issues or using Discourse to socialize an idea before starting the implementation is generally preferable.
```

wjwwood commented on source/Contributing/Developer-Guide.rst:
```diff
@@ -142,6 +149,124 @@ C++ specific
+2. You are implementing a new feature or would like to contribute to OSRF-owned infrastructure (like Jenkins CI)
+
+  * Design doc is required and should be contributed to `ros2/design <https://github.com/ros2/design/>`__ to be made accessible on http://design.ros2.org/
+  * You should fork the repository and submit a pull request detailing the design.
```
nitpick: Try to be consistent on the user or non-use of complete sentences. This is a complete (and punctuated) sentence, but other peer bullet points are not.

wjwwood replied:
There are cases where making a branch on the repository (rather than on a fork) is better. If you have the ability to do so, it actually makes running our CI much easier, though it would be great if at some point that were not the case.

thomas-moulard commented on source/Contributing.rst:
```diff
@@ -6,6 +6,26 @@ Contributing to ROS 2
+
+  OSRF acts as a gate-keeper and advocate for the ROS community. Rely on their expertise and technical judgement from the design phase. Explicit approval from OSRF is required for any non-trivial work.
+
+* Don't fork unnecessarily, adopt community practices whenever possible
```
This one is just for our team too. We can drop it from this PR IMHO (unless you find a way to rephrase it in a more generic way?)

thomas-moulard commented on source/Contributing.rst:
```diff
@@ -6,6 +6,26 @@ Contributing to ROS 2
+
+  Think about your end-customer's experience when developing and contributing. Avoid using non-standard tools or libraries that may not be accessible for everyone.
+
+* Think like ROS product owners
```
`Think like ROS product owners` => `Think about the whole community` may be clearer?

wjwwood commented on source/Contributing/Developer-Guide.rst:
```diff
@@ -142,6 +149,124 @@ C++ specific
+  * Design doc is required and should be contributed to `ros2/design <https://github.com/ros2/design/>`__ to be made accessible on http://design.ros2.org/
+  * You should fork the repository and submit a pull request detailing the design.
+
+  The commit in the pull request must include a link to the ros2/ros2 issue (for example, ``Design doc for task ros2/ros2#<issue id>``).
```
I don't think this is necessary, so long as it is mentioned in the pull request.

thomas-moulard commented on source/Contributing/Developer-Guide.rst:
```diff
@@ -54,13 +54,18 @@ Pull requests
 * Always run CI jobs for all platforms for every pull request and include links to jobs in the pull request.
   (If you don't have access to the Jenkins job someone will trigger the jobs for you.)
+
+* A pull request must ideally contain a single commit.
```
After the doc was published, we realized that it was too extreme. What about:

```
A pull request must contain the minimum number of meaningful commits. `git commit --amend` and `git rebase --interactive` may be used to fix issues in past commits.
```

wjwwood commented on source/Contributing/Developer-Guide.rst:
```diff
@@ -142,6 +149,124 @@ C++ specific
+
+  **After the meeting**
+
+  * The task owner should snd back meeting notes to all attendees
```
`snd` -> `send`

thomas-moulard commented on source/Contributing/Developer-Guide.rst:
```diff
@@ -142,6 +149,124 @@ C++ specific
+
+  * Send a meeting invite at least one week in advance
+  * Meeting duration of one hour is recommended
+  * Booking a meeting room with video conferencing capabilities is recommended
```
This line is Amazon specific, we should drop this line - maybe we could add somewhere: "let's use video when possible because it helps ensuring everyone in the room can speak + it's easier for contributors who are not native English speakers to follow the meeting"

wjwwood commented on source/Contributing/Developer-Guide.rst:
```diff
@@ -142,6 +149,124 @@ C++ specific
+
+Implementation
+^^^^^^^^^^^^^^
+Before starting writing code, learn how to `setup the development environment <https://w.amazon.com/bin/view/AWS_B9/Platform/ROS2_Workspace_Setup/>`__
```
It would be best if this document only linked to other pages within this site, unless absolutely necessary. Perhaps this page could be (re)created in this repository or an existing page could be improved.

thomas-moulard replied:
Remove - Amazon specific

wjwwood commented on source/Contributing/Developer-Guide.rst:
```diff
@@ -142,6 +149,124 @@ C++ specific
+Before starting writing code, learn how to `setup the development environment <https://w.amazon.com/bin/view/AWS_B9/Platform/ROS2_Workspace_Setup/>`__
+and go through `pull request best practices <https://index.ros.org/doc/ros2/Contributing/Developer-Guide/#pull-requests>`_.
+
+* For each repo to modified:
```
`to` -> `to be`

thomas-moulard commented on source/Contributing/Developer-Guide.rst:
```diff
@@ -142,6 +149,124 @@ C++ specific
+
+* Open a pull request for each modified repository
+
+  * Remember to follow `pull request best practices <https://index.ros.org/doc/ros2/Contributing/Developer-Guide/#pull-requests>`_
```
This is this page :)

thomas-moulard commented on source/Contributing/Developer-Guide.rst:
```diff
@@ -142,6 +149,124 @@ C++ specific
+  * `hub <https://hub.github.com/>`__ can be used to create pull requests from the command line
+  * If the task is planned to be released with a specific version of ROS, this information should be included in each pull request
+
+* OSRF engineers who reviewed the design document should be set as reviewers for the pull request
```
Let's change this, we realized that as we don't have write-access to repos this is impossible.
`Engineers who reviewed the design doc should be mentioned in the pull request`

thomas-moulard commented on source/Contributing/Developer-Guide.rst:
```diff
@@ -142,6 +149,124 @@ C++ specific
+  * If the task is planned to be released with a specific version of ROS, this information should be included in each pull request
+
+* OSRF engineers who reviewed the design document should be set as reviewers for the pull request
+* Code review SLO: reviewers should review pull requests within a week and code authors should reply to comments within a week
```
Let's add "best effort" here, this is not internal guidelines anymore so we need to phrase that in a more "open" way.

wjwwood commented on source/Contributing/Developer-Guide.rst:
```diff
@@ -142,6 +149,124 @@ C++ specific
+
+    * A pull request should contain minimal semantically meaningful commits (for instance, a large number of 1-line commits is not acceptable).
+      Already existing commits can be amended to avoid creating a new commit every time, using ``git commit --amend``
+    * Each commit must have a properly written, meaningful, commit message. More instructions `here <https://chris.beams.io/posts/git-commit/>`__.
```
Again, this is fine, but if the commits have already been pushed, amending them will require a force push and can be disruptive to other collaborators, reviewers, and/or testers.

thomas-moulard commented on source/Contributing/Developer-Guide.rst:
```diff
@@ -142,6 +149,124 @@ C++ specific
+
+* OSRF engineers who reviewed the design document should be set as reviewers for the pull request
+* Code review SLO: reviewers should review pull requests within a week and code authors should reply to comments within a week
+* Iterate on feedback as usual, amend and ``push -f`` to update the development branch as needed
```
We don't need `push -f` anymore as we don't enforce the single commit rule.

wjwwood replied:
Again, avoiding `push -f` (or the better alternative `push --force-with-lease`) should be avoided in my opinion.

wjwwood commented on source/Contributing/Developer-Guide.rst:
```diff
@@ -142,6 +149,124 @@ C++ specific
+      Already existing commits can be amended to avoid creating a new commit every time, using ``git commit --amend``
+    * Each commit must have a properly written, meaningful, commit message. More instructions `here <https://chris.beams.io/posts/git-commit/>`__.
+    * Moving files must be done in a separate commit, otherwise git may fail to accurately track the file history
+    * The last commit message must contain a reference to the ros2/ros2 issue to ensure that merging the PR will close the issue.
```
This can be accomplished with a comment in the pull request description, so it is not necessary for the commits to contain them.

thomas-moulard commented on source/Contributing/Developer-Guide.rst:
```diff
@@ -142,6 +149,124 @@ C++ specific
+* OSRF engineers who reviewed the design document should be set as reviewers for the pull request
+* Code review SLO: reviewers should review pull requests within a week and code authors should reply to comments within a week
+* Iterate on feedback as usual, amend and ``push -f`` to update the development branch as needed
+* Once the PR is approved, OSRF engineers will merge it in
```
This sentence is not finished?
Additionally, it's not generally correct to say `OSRF engineers` there (it is for us because we just contribute to RMW/RCL for now). Maybe something like `Package maintainers` wil...

wjwwood replied:
` in` -> `.`

juanrh commented on source/Contributing/Developer-Guide.rst:
```diff
@@ -142,6 +149,124 @@ C++ specific
+
+  * Modify the code. Go to next step if finished or at regular interval to backup your work.
+  * `Self review <https://git-scm.com/book/en/v2/Git-Tools-Interactive-Staging>`__ your changes using ``git add -i``
+  * Create a new commit using ``git commit -s``
```
Is [code signing](https://help.github.com/en/articles/signing-commits) also required, besides sign off?

wjwwood replied:
It's a good idea, but it's not required for contributing to ROS 2 right now. The only reason we require sign-off is to satisfy the DCO requirements/CI bot.

dirk-thomas commented on source/Contributing/Developer-Guide.rst:
```diff
@@ -86,6 +91,8 @@ Development Process
   Besides using automated tests, also run the modified code path manually to ensure that the patch works as intended.
 * Always run CI jobs for all platforms for every pull request and include links to the jobs in the pull request.
 
+For more details on recommended software development workflow, see `software development lifecycle <https://index.ros.org/doc/ros2/Contributing/Developer-Guide/#software-development-lifecycle>`_ section.
```
Please use relative links instead of absolute ones.

dirk-thomas commented on source/Contributing/Developer-Guide.rst:
```diff
@@ -142,6 +149,124 @@ C++ specific
+Task creation
+^^^^^^^^^^^^^
+Tasks requiring changes to critical parts of ROS 2 should have their design scheduled during the first half of the release cycle.
+If a design review is happening in the second half of the release cycle, it should be targeted toward the N+2 release.
```
`N+2` might be technically correct under some assumption what `N` is but I would suggest to rephrase this to describe this in text form instead.

dirk-thomas commented on source/Contributing/Developer-Guide.rst:
```diff
@@ -142,6 +149,124 @@ C++ specific
+If a design review is happening in the second half of the release cycle, it should be targeted toward the N+2 release.
+In case an exception is necessary, OSRF leaders should be notified in advanced and a clear path should be provided (for example, what type of changes will be tolerated in patch releases).
+
+* An issue should be created in the `ros2/ros2 <https://github.com/ros2/ros2/issues>`__ org, clearly describing the task being worked on.
```
Not every ticket needs to be in the `ros2/ros2` repo. If the change is scoped to another repo it can and should be raised there.

prajakta-gokhale commented:
### Revision 2
* Removed Amazon specific bits, made some parts more generic
* Fixed typos and grammatical errors
* Used relative links, fixed incorrect and obscure steps

@thomas-moulard @wjwwood please take a look and let me know if any changes are needed.

wjwwood commented on source/Contributing/Developer-Guide.rst:
```diff
@@ -142,6 +149,122 @@ C++ specific
+  * You should fork the repository and submit a pull request detailing the design.
+
+  Mention the related ros2 issue (for example, ``Design doc for task ros2/ros2#<issue id>``) in the pull request or the commit message.
+  Detailed instructions are on the `ROS2 Contribute <http://design.ros2.org/contribute.html>`__ page. Design comments will made directly on the pull request.
```
nitpick: one sentence per line

wjwwood commented on source/Contributing/Developer-Guide.rst:
```diff
@@ -142,6 +149,122 @@ C++ specific
+
+Design document review
+^^^^^^^^^^^^^^^^^^^^^^
+Once the design is ready for review, a pull request should be opened and appropriate reviewers should be assigned.
```
nitpick: blank line between heading and content

thomas-moulard commented on source/Contributing.rst:
```diff
@@ -6,6 +6,34 @@ Contributing to ROS 2
+
+* Adopt community best-practices whenever possible instead of ad-hoc processes
+
+  Think about your end-customer's experience when developing and contributing.
```
end-user?

thomas-moulard commented on source/Contributing.rst:
```diff
@@ -6,6 +6,34 @@ Contributing to ROS 2
+* Think about the community as a whole
+
+  Be mindful of the whole community.
+  You may have to interact with other contributors in the future and it is important to build trust with everyone from early on.
```
For this one, the original intent was more around this:
"Think of the bigger picture. Other developers are building different robots with different constraints. ROS needs to accommodate the requirements of the whole community."

We may want to keep something around those lines.

prajakta-gokhale commented:
### Revision 3
* Addressed comments
* Fixed new markdown linter warnings

prajakta-gokhale commented:
@wjwwood please see if this is ready to merge. Thanks!

wjwwood commented:
@prajakta-gokhale you're good to go, my previous approval stands since the nitpicks were addressed. 👍

wjwwood commented:
I assume the Travis CI is sufficient. @dirk-thomas are your comments addressed?

dirk-thomas commented on source/Contributing/Developer-Guide.rst:
```diff
@@ -86,6 +91,8 @@ Development Process
   Besides using automated tests, also run the modified code path manually to ensure that the patch works as intended.
 * Always run CI jobs for all platforms for every pull request and include links to the jobs in the pull request.
 
+For more details on recommended software development workflow, see `Software Development Lifecycle`_ section.
```
Does this link actually work?

prajakta-gokhale replied:
Yes, verified that it works.

dirk-thomas commented on source/Contributing/Developer-Guide.rst:
```diff
@@ -142,6 +149,127 @@ C++ specific
+
+Tasks requiring changes to critical parts of ROS 2 should have design reviews during early stages of the release cycle.
+If a design review is happening in the later stages, the changes will be part of a future release.
+In case an exception is necessary, OSRF leaders should be notified in advanced and a clear path should be provided (for example, what type of changes will be tolerated in patch releases).
```
Requiring to contact "OSRF leaders" seems off to me. Since I can't come up with a sane formulation of this sentence I would suggest to remove it.

dirk-thomas commented on source/Contributing/Developer-Guide.rst:
```diff
@@ -142,6 +149,127 @@ C++ specific
+
+  * Send a meeting invite at least one week in advance
+  * Meeting duration of one hour is recommended
+  * Meeting invite should list all decisions to be made during the review (decisions requiring OSRF approval)
```
To phrase this more generically I would say decision require *maintainer* approval. There are already higher level packages which OSRF has no "control" over and isn't the right gatekeeper.

The same might apply in other parts where OSRF is mentioned explicitly.

dirk-thomas commented on source/Contributing/Developer-Guide.rst:
```diff
@@ -142,6 +149,127 @@ C++ specific
+Implementation
+^^^^^^^^^^^^^^
+
+Before starting, go through `pull request best practices <pull-requests>`_.
```
Same question for this link?

Same below.

prajakta-gokhale replied:
Good catch - modified.

hidmic commented:
@prajakta-gokhale will you be able to keep moving this PR forward?

prajakta-gokhale commented:
@hidmic yes I'll be working on it.

prajakta-gokhale commented:
@dirk-thomas could you please take a look at the new revision that has your feedback addressed?

dirk-thomas commented:
Thank you for the improvements.
